lib/std/mem/Allocator.zig

```diff
@@ -318,6 +318,13 @@ pub fn dupeZ(allocator: Allocator, comptime T: type, m: []const T) ![:0]T {
     return new_buf[0..m.len :0];
 }
 
+inline fn ensureSlice(comptime Type: type, comptime function_name: []const u8) std.builtin.Type.Pointer {
+    return switch (@typeInfo(Type)) {
+        .Pointer => |pointer| pointer,
+        else => @compileError(function_name ++ " expects an array but received value of type `" ++ @typeName(Type) ++ "`"),
```

Comment:
The terms "slice", "array", and "pointer" are all used pretty interchangably here. Is that valid?

Reply:
I think "slice" would be the more appropriate term here. Although slices are essentially just references to arrays, in zig, arrays are treated as first-class values, and so "array" refers to a value whose type is of the form `[n]T`, which would not be the correct type to pass.

This also makes me wonder whether `free` should accept `*[n]T`. Although that can coerce to `[]T`, someone who's trying to free that most likely allocated it using `create` instead of `alloc`, or sliced using comptime-known indices, which also might indicate that they are erroneously trying to free only a segment of an allocation, which would violate the API.

So, if you want to, you could consider adding a parameter like `size: std.builtin.Type.Pointer.Size` to `ensureSlice`, and check that it matches `pointer.size`, and then do `ensureSlice(@TypeOf(ptr), .One)` in `destroy`, and vice versa in `free` and `resize`. And with that, you could also make the error's terminology depend on the expected pointer size, e.g.
```zig
const expected_str = switch (size) {
    .One => " a single item pointer ",
    .Slice => " a slice ",
    else => unreachable,
};
// -- snip --
@compileError(std.fmt.comptimePrint("{s} expects {s} but received a value of type `{s}`", .{ function_name, @typeName(Type) }));
```

Reply:
I'm working on implementing this, and I've got a couple of questions:
1. Do think this restriction is likely to break any valid code? I'll run the standard library tests to maybe that'll give some good assurance about "real" code.
2. What size argument should I give for `realloc` and `reallocAdvanced`?

Reply:
For those, I think the only logical choice is `.Slice`, being that it doesn't quite make sense to try and resize a `*u32`, or a `*[2]u16`

Reply:
As for breaking valid code, I imagine there may be some edge cases, though I imagine they'd be quite trivial. And like I said, in the best case, it might even uncover some correctness errors.

Reply:
I've implemented that here, but it breaks the tests (possibly in a good way!): https://github.com/leoconst/zig/pull/1

The issue is, running `zig build test-std -Dskip-release -Dskip-non-native -freference-trace` now gives me:
```
error: free__anon_65422 expects a slice but received a value of type `*align(4096) [4096]u8`
    @compileError(std.fmt.comptimePrint("{s} expects {s} but received a value of type `{s}`", .{ function_name, expectation, @typeName(Type) }));
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lib\std\mem\Allocator.zig:296:30: note: called from here
    const Slice = ensureSlice(@TypeOf(memory), @src().fn_name, .Slice);
                  ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
error: test...
```
Unfortunately, it doesn't really tell me where this error is coming from, so I'm a bit stuck.

Also, I've just realised my [most recent commit](https://github.com/ziglang/zig/pull/14520/commits/cdb210a8a28f0a8a7aace8cdbe48ab4a1a38ec62) has regressed the name from `free` to `free__anon_65422`! Maybe I should revert?

Thank you for all of your help so far by the way :-)

Reply:
oof, that's unfortunate, but I suppose it makes sense. It's a generic function, so the compiler annotates its name as such. So yeah, best to do it manually.

At any rate, to track down where it's coming from, my immediate (though mayhaps naive) approach would be to just trial and error it by `zig test`ing specific files, so you can narrow down where it can and where it can't be happening.

Reply:
Okey dokey. I've reverted the `@src()` usage. I'll have a poke around to see if I can find the culprit but there's a high likelihood I'll have to call in some help!

Reply:
I suppose one this worth asking is - why is the user not shown the callsite? That would be a desirable feature, and it's something I mentioned in the description as missing when calling the example. Is it a limitation of `@compileError`? Is there anything I can/should do about it?

Reply:
Afaik the situation regarding error locations is just a bit finicky, I don't think it's something that's in scope for this PR.